j5/backends/hardware/sr/v4/motor_board.py

```diff
@@ -0,0 +1,190 @@
+"""Hardware Backend for the SR v4 motor board."""
+from functools import wraps
+from typing import Callable, List, Optional, Type, TypeVar, cast
+
+from serial import Serial, SerialException, SerialTimeoutException
+from serial.tools.list_ports import comports
+from serial.tools.list_ports_common import ListPortInfo
+
+from j5.backends import Backend, CommunicationError
+from j5.backends.hardware.env import HardwareEnvironment
+from j5.boards import Board
+from j5.boards.sr.v4.motor_board import MotorBoard
+from j5.components.motor import MotorInterface, MotorSpecialState, MotorState
+
+CMD_RESET = 0
+CMD_VERSION = 1
+CMD_MOTOR = [2, 3]
+CMD_BOOTLOADER = 4
+
+SPEED_COAST = 1
+SPEED_BRAKE = 2
+
+RT = TypeVar("RT")  # pragma: nocover
+
+
+def handle_serial_error(func: Callable[..., RT]) -> Callable[..., RT]:  # type: ignore
+    """
+    Wrap functions that use the serial port, and rethrow the errors.
+
+    This is a decorator that should be used to wrap any functions that call the serial
+    interface. It will catch and rethrow the errors as a CommunicationError, so that it
+    is more explicit what is going wrong.
+    """
+    @wraps(func)
+    def catch_exceptions(*args, **kwargs):  # type: ignore
+        try:
+            return func(*args, **kwargs)
+        except SerialTimeoutException as e:
+            raise CommunicationError(f"Serial Timeout Error: {e}")
+        except SerialException as e:
+            raise CommunicationError(f"Serial Error: {e}")
+    return catch_exceptions
+
+
+class SRV4MotorBoardHardwareBackend(
+    MotorInterface,
+    Backend,
+):
+    """The hardware implementation of the SR v4 motor board."""
+
+    environment = HardwareEnvironment
+    board = MotorBoard
+
+    @classmethod
+    def discover(
+            cls,
+            find: Callable = comports,
+            serial_class: Type[Serial] = Serial,
+    ) -> List[Board]:
+        """Discover all connected motor boards."""
+        # Find all serial ports.
+        ports: List[ListPortInfo] = find()
+
+        # Filter down to just motor boards.
+        ports = list(filter(lambda x: x.vid == 0x403, ports))  # FTDI, Ltd
```

Comment:
Instead of having a ton of `ports = list(filter(`, you could have a function defined here that does all the checks, and then have a single filter

Reply:
Or you could have a dictionary that states what each of the attrs needs to be, then iterate over that with an `all`

Reply:
I wrote a bunch of stuff about this before I saw @Skybound1's comment, so here's the interesting bit:

```python
def is_motor_board(port):
  return port.manufacturer == "Student Robotics" and port.product == "MCV4B" \
    and port.vid == 0x0403 and port.pid == 0x6001
...
for port in filter(is_motor_board, comports()):
  ...
```

Using `all` and a dictionary is a nice approach, though I personally think its overly complex for this use.

Reply:
I was thinking something along the lines of 

```
parameters = {
    'manufacturer': 'Student Robotics',
    'product': 'MCV48,
...
}
for port in comports:
    if all(getattr(port, k) == v for k, v in parameters.keys():
```

you this is getting too crazy xD

Reply:
For this pull request, I'll go with @kierdavis's approach, but if we refactor to have a generic serial device backend in the future, I can see the merits of @Skybound1's approach.

Reply:
46c3740